scipy/spatial/transform/_rigid_transformation.pyx

```diff
@@ -0,0 +1,1979 @@
+
+        Unit dual quaternions encode orientation in a real unit quaternion
+        and translation in a dual quaternion. There is a double cover, i.e.,
+        the unit dual quaternions q and -q represent the same transformation.
+
+        Unit dual quaternions must have a real quaternion with unit norm and
+        a dual quaternion that is orthogonal to the real quaternion to satisfy
+        the unit norm constraint. This function will enforce both properties
+        through normalization.
+
+        Parameters
+        ----------
+        dual_quat : array_like, shape (N, 8) or (8,)
+            A single unit dual quaternion or a stack of unit dual quaternions.
+            The real part is stored in the first four components and the dual
+            part in the last four components.
+        scalar_first : bool, optional
+            Whether the scalar component goes first or last in the two
+            individual quaternions that represent the real and the dual part.
+            Default is False, i.e. the scalar-last order is used.
+
+        Returns
+        -------
+        transformation : `RigidTransformation` instance
+            A single transformation or a stack of transformations.
+
+        Examples
+        --------
+        >>> from scipy.spatial.transform import RigidTransformation as Tf
+        >>> import numpy as np
+
+        Creating from a single unit dual quaternion:
+
+        >>> tf = Tf.from_dual_quat([
+        ...     0.0617101, -0.06483886, 0.31432811, 0.94508498,
+        ...     0.04985168, -0.26119618, 0.1691491, -0.07743254])
+        >>> tf.as_matrix()
+        array([[0.79398752, -0.60213598, -0.08376202, 0.24605262],
+               [0.58613113, 0.79477941, -0.15740392, -0.4932833],
+               [0.16135089, 0.07588122, 0.98397557, 0.34262676],
+               [0., 0., 0., 1.]])
+        >>> tf.single
+        True
+        """
+        dual_quat = np.asarray(dual_quat, dtype=float)
+
+        if (dual_quat.ndim not in [1, 2] or dual_quat.shape[0] == 0
+                or dual_quat.shape[-1] != 8):
+            raise ValueError(
+                "Expected `dual_quat` to have shape (8,), or (N, 8), "
+                f"got {dual_quat.shape}.")
+
+        single = dual_quat.ndim == 1
+        dual_quat = np.atleast_2d(dual_quat)
+
+        real_part = dual_quat[:, :4]
+        dual_part = dual_quat[:, 4:]
+        if scalar_first:
+            real_part = np.roll(real_part, -1, axis=1)
+            dual_part = np.roll(dual_part, -1, axis=1)
+
+        real_part, dual_part = _normalize_dual_quaternion(real_part, dual_part)
+
+        matrix = np.empty((len(dual_quat), 4, 4), dtype=float)
+        rotation = Rotation.from_quat(real_part)
+
+        matrix[:, :3, :3] = rotation.as_matrix()
+        matrix[:, :3, 3] = 2.0 * np.asarray(
+            compose_quat(dual_part, rotation.inv().as_quat()))[:, :3]
+        matrix[:, 3, :3] = 0.0
+        matrix[:, 3, 3] = 1.0
```

Comment:
As a small improvement put such construction block into the function or private method. Can be reused at leas for `from_exp_coords`.

Reply:
Good idea, moved to a private method.